requirements.txt

```
@@ -3,3 +3,5 @@ torch
 six
 regex
 numpy
+transformers
+deepspeed
```

We need a  very specific version of the deepspeed branch:

```suggestion
git+https://github.com/microsoft/DeepSpeed.git@big-science
```

---

the trouble with that is that it will rebuild it every time requirements are run, since it won't know which version was built. but I guess it's better than how it was before.

---

If we end up specifying `deepspeed` in this fashion, should we also update the current README?

---

It's already so: https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/README.md#setup

It's just that in the readme instructions it's easier to be demonstrative with explicit branching, which we can't do in the `requirements` file.

---

Right, I meant specifying `deepspeed` in requirements might add more confusion since README already covers the installation process with explicit branching. So perhaps it's better to leave `deepspeed` out of requirements all together...

---

either way works for me, or perhaps add it to the file, but commented out? So it'll be easy to enable for someone who wants it that way?

---

Oh, commenting out sounds like a good idea. Not sure if it's worth a whole new PR, I'll push it up to main if that's fine with you? Below is what the updated `requirements.txt` would look like. I might update README to indicate that uncommenting this line would also install the required version of DeepSpeed.

```
pybind11
torch
six
regex
numpy
transformers
# git+https://github.com/microsoft/DeepSpeed.git@big-science
```

---

Sounds good, Jake

> I might update README to indicate that uncommenting this line would also install the required version of DeepSpeed.

OK.

Though pre-building is better, as it leads to less noise at run time. but that's a personal preference.

---

Another idea that just came to my mind is to use git submodules, so that when people clone the repo, they can simply do 

```
git clone --recurse-submodules https://github.com/bigscience-workshop/Megatron-DeepSpeed.git
```

and it will clone the specific branch of DeepSpeed automatically. We can do the same for apex. If we really want to, we can have an `install.sh` that runs the sequence of commands to install all three (Megatron-DeepSpeed, apex, and DeepSpeed) in one go.

---

I personally am not a fan of submodules as it makes things more complicated. For example you can't do a simple `git pull` and have to do it in a special way with submodules. 

Moreover it leads to multiple entry points, since I for example use source clones of say deepspeed for other projects. And it becomes a mess trying to manage clones that aren't aligned under the same dir.

I think `git+https://github.com/microsoft/DeepSpeed.git@big-science` in `requirement.txt` is a much simpler solution for those who want things automatic. Especially since we hope it won't change much at all for our needs.

The install instructions are copy-n-paste too, so they are pretty automated as well.

---

apex, you don't really need to change either. It's very stable and in fact is planned to be phased out.

---

That's said an optional `install.sh` is surely an option if it saves you and others time and simplifies things.

---

I guess there are more complexities to this than I had thought. I agree that it's easy to complicate things once submodules are introduced. Appreciate your feedback!